src/ngraph/runtime/executable.hpp

```diff
@@ -51,6 +54,16 @@ class ngraph::runtime::Executable
     bool call_with_validate(const std::vector<std::shared_ptr<runtime::Tensor>>& outputs,
                             const std::vector<std::shared_ptr<runtime::Tensor>>& inputs);
 
+    /// \brief Asynchronously executes a single iteration of the Function. The `future` is used
+    ///     to monitor the execution.
+    /// \param outputs vector of runtime::Tensor used as outputs
+    /// \param inputs vector of runtime::Tensor used as inputs
+    /// \returns a valid std::future to monitor the execution. Use future.get() to get the results
+    ///     or future.wait*() to wait for completion.
+    virtual std::future<void>
```

Comment:
This is the classic problem with `std::future<>` -- there's no `.then()` method, so you either have to poll or burn a thread waiting.  :-/

It might be good to return a different object, or take a callback.

Reply:
It seems to me that doing a callback only pushes the issue down one level. There is a new thread created which does the callback and that thread would be waiting on some future-like thing. The future is the stl way of doing this.